observability/akka-http-metrics/src/test/suite/scala/com/daml/metrics/akkahttp/WebSocketMetricsSpec.scala

```diff
@@ -0,0 +1,342 @@
+        }
+    }
+  }
+
+  // drain a source, and return the content of each message
+  private val drainStreamedMessages: Sink[Message, Future[List[List[ByteString]]]] =
+    Flow[Message]
+      .flatMapConcat {
+        case TextMessage.Strict(text) =>
+          Source.single(List(ByteString(text)))
+        case TextMessage.Streamed(textStream) =>
+          textStream.fold(List[ByteString]())((acc, c) => acc :+ ByteString(c))
+        case BinaryMessage.Strict(data) =>
+          Source.single(List(data))
+        case BinaryMessage.Streamed(dataStream) =>
+          // dataStream.map(_ => 1)
+          dataStream.fold(List[ByteString]())((acc, c) => acc :+ c)
+      }
+      .toMat(Sink.fold(List[List[ByteString]]())((acc, s) => acc :+ s))((_, mat2) => mat2)
+
+  private val getMessagesReceivedTotalValue: TestMetrics => Long = (metrics: TestMetrics) =>
+    metrics.messagesReceivedTotalValue
+
+  private val getMessagesSentTotalValue: TestMetrics => Long = (metrics: TestMetrics) =>
+    metrics.messagesSentTotalValue
+
+  private val getMessagesReceivedBytesTotalValue: TestMetrics => Long =
+    (metrics: TestMetrics) => metrics.messagesReceivedBytesTotalValue
+
+  private val getMessagesSentBytesTotalValue: TestMetrics => Long = (metrics: TestMetrics) =>
+    metrics.messagesSentBytesTotalValue
+
+  private def checkCountThroughFlow(
+      messages: List[Message],
+      metricExtractor: (TestMetrics) => Long,
+      expected: Long,
+  ): Future[Assertion] = {
+    withDuplicatingFlowAndMetrics { (flow, metrics) =>
+      val done = Source(messages).via(flow).runWith(drainStreamedMessages)
+      done.map { _ =>
+        metricExtractor(metrics) should be(expected)
+      }
+    }
+  }
+
+  private def checkCountThroughFlow(
+      message: Message,
+      metricExtractor: TestMetrics => Long,
+      expected: Long,
+  ): Future[Assertion] = {
+    checkCountThroughFlow(List(message), metricExtractor, expected)
+  }
+
+  "websocket_message_received_total" should {
+    "count passing strict text messages" in {
+      checkCountThroughFlow(strictTextMessage, getMessagesReceivedTotalValue, 1)
+    }
+
+    "count passing streamed text messages" in {
+      checkCountThroughFlow(streamedTextMessage, getMessagesReceivedTotalValue, 1)
+    }
+
+    "count passing strict binary messages" in {
+      checkCountThroughFlow(strictBinaryMessage, getMessagesReceivedTotalValue, 1)
+    }
+
+    "count passing streamed binary messages" in {
+      checkCountThroughFlow(streamedBinaryMessage, getMessagesReceivedTotalValue, 1)
+    }
+
+    "count a set of passing messages" in {
+      checkCountThroughFlow(
+        List[Message](
+          strictTextMessage,
+          streamedTextMessage,
+          strictBinaryMessage,
+          streamedBinaryMessage,
+        ),
+        getMessagesReceivedTotalValue,
+        4,
+      )
+    }
+  }
+
+  "websocket_message_sent_total" should {
+    "count passing strict text messages" in {
```

Comment:
I don't think it fits well in this case.
It is not checking properties, it is checking the results of an operation.
Those tests will be updated when adding labels to the metrics. I'll look at it then.

Reply:
You can use the above suggestion not just for properties, i.e., it's just a method to make test cases easier to read in case you just want to run the same operation / function with different combination of input parameters. In this particular case, you could group test cases per metric. E.g., declare as input parameter the type of the message and declare one row for each message input type (`strictTextMessage`, `streamedTextMessage`, etc.) then check one metric for all of them (and then this could be repeated for the other metrics as well).
I'm OK with keeping the test cases as-is.

Reply:
I tested it, it is less bad that I was thinking. But I have 2 problems:
* it stops at the first the failure. So it can hide the fact that there is more than on problem.
* the output gets quite complex, because the inputs are not simple types. Finding out exactly which fails is a bit more complex:
  ```
  TestFailedException was thrown during property evaluation. (WebSocketMetricsSpec.scala:75)
    Message: 1 was not equal to 2
    Location: (WebSocketMetricsSpec.scala:277)
    Occurred at table row 1 (zero based, not counting headings), which had values (
      input = List(TextMessage.Streamed(Source(SourceShape(StatefulMapConcat.out(730583371))))),
      expectedCount = 2
    )
  ```